### PR TITLE
add(blog): exports, export default and export

### PR DESCRIPTION
**Pull request by USTC-Han**
添加 exports、 export default and export之间的关系

### CODE REVIEW

File: blog/2019-12-05-exports-vs-export.md [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+我们知道浏览器是不支持ES6模块语法的，所以对于代码中的ES6模块一般是采用Babel转成CommonJS规范的代码。此时的代码虽然浏览器仍然不能执行，但是Node可以，因为webpack是基于Node构建的，所以我们可以通过webpack将代码打包成浏览器支持的ES5代码。
+
+<!--truncate-->
+
+<Img width="480" legend="图：ES6模块转浏览器可执行代码示意图" src="https://cosmos-x.oss-cn-hangzhou.aliyuncs.com/y8Gvm6.png" />
+
+为了探索ES6模块中`export`和`export default`这两个之间的关系，我们可以看下面这个例子
+
+```js
+  export var info = {age: 22, name: 'han'}
```
Comment (muwenzi) [inline blog/2019-12-05-exports-vs-export.md]: indent

File: blog/2019-12-05-exports-vs-export.md [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+  var email = 'SA17225105@gmail.com'
```
Comment (muwenzi) [inline blog/2019-12-05-exports-vs-export.md]: personal email will be sensitive on Internet and suggest using fake email instead

File: blog/2019-12-05-exports-vs-export.md [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+console.log(c.count);
+```
+
+通过上面的两个例子可以看出，例1中输出的是值的拷贝（commonJS模块），而例2中使用`getter`了，当外部获取`c.count`时，会去动态调取函数中最新的值（ES6模块）。
+
+如果上面的例子明白的话，那么**CommonJS 模块输出的是一个值的拷贝，ES6 模块输出的是值的引用**这句话也不难理解。在commonJS模块就是采用`{key: value}`的形式传给导入模块，传递的是本模块导出对象的拷贝。而ES6模块中则是对每个导出的变量都设置了`getter`，外部的导入模块取值时都是执行对应的`getter`函数，然后返回执行过的值。
+
+### 运行时加载 VS 静态加载
+
+通过上面的分析，我们知道CommonJS模块导入的时候，导入的是拷贝的一个对象，导入后便于该模块无关了。所以当执行到`require（）`的时候，会执行导入模块一遍（无缓存），然后将执行后的对象拷贝到当前文件，然后文件执行时才能取到导入对象里的元素。在导入语句未执行时，是不知道具体要导入的是什么的。
```
Comment (muwenzi) [inline blog/2019-12-05-exports-vs-export.md]: `require（）` -> `require()`

File: blog/2019-12-05-exports-vs-export.md [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+`export`和`exports`属于不同的规范，显然不是一回事。如果在项目中导出模块混用`export`和`exports`，模块在转义和打包的时候会将里面的代码统统打包成浏览器支持运行的ES5代码。
+
+但是还是不建议混写两者，一方面容易使人懵逼，另一方面两个规范的输出规则是完全不一样的。
+
+## ES6 模块与 CommonJS 模块的差异
+
+> 1. **CommonJS 模块输出的是一个值的拷贝，ES6 模块输出的是值的引用**
+> 2. **CommonJS 模块是运行时加载，ES6 模块是静态加载。**
+> 3. **ES6 模块的运行环境是严格模式**
```
Comment (muwenzi) [inline blog/2019-12-05-exports-vs-export.md]: There is also the difference between synchronous and asynchronous

File: blog/2019-12-05-exports-vs-export.md (hunk 2 above)
Comment (muwenzi) [reply blog/2019-12-05-exports-vs-export.md]: indent

File: blog/2019-12-05-exports-vs-export.md (hunk 4 above)
Comment (USTC-Han) [reply blog/2019-12-05-exports-vs-export.md]: ES6  is asynchronous？

File: docs/javascript/6.modules/es6-module.md [hunk 5]
```diff
@@ -305,9 +315,10 @@ var a = require('a');
 
 对于`CommonJS`而言循环引用的时候会去执行引用的模块，并输出引用模块已经执行部分`exports`的值，还未执行的部分不会输出。
 
-对于`ES6`的`import`而言，执行到`import`的时候，它不会去执行引入模块的代码，而只是返回模块的引用，只有在真正调用引用模块中的值的时候才会执行。也就是说`ES6`的本身不会关注代码是否在循环引用，这需要编写代码的人去保证在真正取值的时候能够取道值，不会出错。
+对于`ES6`的`import`而言，执行到`import`的时候，它不会去执行引入模块的代码，而只是返回模块的引用，只有在真正调用引用模块中的值的时候才会执行。也就是说`ES6`的本身不会关注代码是否在循环引用，这需要编写代码的人去保证在真正取值的时候能够取到值，不会出错。
```
Comment (muwenzi) [inline docs/javascript/6.modules/es6-module.md]: I’m a bit confused, run this example https://exploringjs.com/es6/ch_modules.html#sec_cyclic-dependencies in node env and find out the diff

In addition, 编写代码的人 a bit weird, 开发者 or ...

File: docs/javascript/6.modules/es6-module.md [hunk 6]
```diff
@@ -303,23 +269,113 @@ console.log(c.count);
...
+//------ b.js ------
+var a = require('a'); //位置2
+function bar() {
+    if (Math.random()) {
+        a.foo(); // 位置3
+    }
+}
+exports.bar = bar;
+```
+当a模块执行到位置1的时候，回去加载b模块。当b模块执行到位置2的时候，去执行a模块，结果什么也不会拿到。此时当执行到位置3时会因为不存在`a.foo()`报错。
```
Comment (muwenzi) [inline docs/javascript/6.modules/es6-module.md]: 回去?

File: docs/javascript/6.modules/es6-module.md [hunk 7]
```diff
@@ -303,23 +269,113 @@ console.log(c.count);
...
+import {foo} from 'a'; // 位置3
+export function bar() {
+    if (Math.random()) {
+        foo(); // 位置4
+    }
+}
+```
+执行a模块到位置1时，导入的变量`bar`和b模块的`bar()`函数已经建立了引用关系，所以执行到位置2的时候是会动态执行b模块中的`bar()`函数，当执行到位置4的时候，会动态执行a模块中的`foo()`函数，此时形成一个循环引用关系。
+
+虽然上面的例子是可以执行的，但是将上述代码直接执行会因为两个模块之间循环引用次数过多直接爆栈。如果想让上面的代码执行，可以加个限制条件，例如`Math.random()> 0.1`。这样可以限制循环引用次数，避免爆栈。
```
Comment (muwenzi) [inline docs/javascript/6.modules/es6-module.md]: `Math.random()> 0.1` cannot entirely avoided ?

Comment (USTC-Han) [reply docs/javascript/6.modules/es6-module.md]: 如果不加限制条件，会爆栈的
test

**Pull request merged**